resources/net.go

```diff
@@ -0,0 +1,671 @@
+	IPv6    string `yaml:"ipv6"`    // ipv6 address in cidr format
+	Gateway string `yaml:"gateway"` // gateway address
+
+	SoftFail bool `yaml:"softfail"` // don't error on connectivity loss
+
+	ipv4 string // the interface's actual ipv4 address
+	ipv6 string // the interface's actual ipv4 address
+
+	unitFile string // the interface unit file path
+
+	conn *systemd.Conn // dbus connection
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *NetRes) Default() Res {
+	return &NetRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+		State: "up",
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *NetRes) Validate() error {
+	// validate state
+	if obj.State != "up" && obj.State != "down" {
+		return fmt.Errorf("invalid state: %s", obj.State)
+	}
+
+	// validate network address input
+	if obj.IPv4 != "" {
+		if _, _, err := net.ParseCIDR(obj.IPv4); err != nil {
+			return errwrap.Wrapf(err, "error parsing ipv4: %s", obj.IPv4)
+		}
+	}
+	if obj.IPv6 != "" {
+		if _, _, err := net.ParseCIDR(obj.IPv6); err != nil {
+			return errwrap.Wrapf(err, "error parsing ipv6: %s", obj.IPv6)
+		}
+	}
+	if obj.Gateway != "" {
+		if g := net.ParseIP(obj.Gateway); g == nil {
+			return fmt.Errorf("could not parse gateway: %s", obj.Gateway)
+		}
+	}
+
+	// validate the interface name
+	_, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+	return obj.BaseRes.Validate()
+}
+
+// Init runs some startup code for this resource.
+func (obj *NetRes) Init() error {
+	// establish a dbus connection
+	var err error
+	obj.conn, err = systemd.New()
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to connect to systemd")
+	}
+
+	return obj.BaseRes.Init()
+}
+
+// Watch listens for networkd events by eavesdropping on the DBus Properties
+// interface (org.freedesktop.DBus.Properties) and filtering for signals from
+// the network device. It also watches the networkd service via a dbus unit
+// subscription, and the unit file responsible for configuring the interface.
+func (obj *NetRes) Watch() error {
+	// make sure systemd is running
+	if !systemdUtil.IsRunningSystemd() {
+		return fmt.Errorf("systemd is not running")
+	}
+
+	// create a dbus subscription to watch the networkd service unit
+	set := obj.conn.NewSubscriptionSet()
+	svcChan, svcErr := set.Subscribe() // svc event channel
+	set.Add(NetworkdService)
+
+	// get the network link's dbus path
+	ifacePath, err := dbusIfacePath(obj.GetName())
+	if err != nil {
+		return errwrap.Wrapf(err, "error getting iface dbus path")
+	}
+
+	// create a private message bus for watching the network device
+	bus, err := util.SystemBusPrivateUsable()
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to connect to bus")
+	}
+	defer bus.Close()
+	// Add a DBus rule to watch the "PropertiesChanged" signal and get messages
```

Comment:
I tried to put the eavesdropping stuff in a function. but I was having trouble handling the bus.Close() and associated operations without closing stuff too early/not closing at all. I'd like to try again, for the sake of unit testing and style, but I need to re-evaluate my approach.